### PR TITLE
Add direct migration filesystem ownership enhancement

### PR DESCRIPTION
**Pull request by eriknelson**
I wanted to capture our understanding of the filesystem ownership problem and proposed solutions in a more permanent location than a Bugzilla comment, so submitting here for review for posterity.

### CODE REVIEW

File: enhancements/direct-migration-filesystem-ownership/README.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+The group, in this case, is the same as the application user, hence the application pod does not break.)
+
+## Proposed solution
+
+Before starting the rsync client pod, we check if the pvc’s are attached to
+any pods in the namespace.
+
+* If any pvc is attached to one or more pods:
+  * Attached to one pod, get its fsGroup and use the same fsGroup for the rsync
```
Comment (shawn-hurley) [inline enhancements/direct-migration-filesystem-ownership/README.md:95]: We should supplemental groups here as well IMO

File: enhancements/direct-migration-filesystem-ownership/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  for the rsync client pod while explicitly setting the node to match the picked pod’s.
+  * If pvc’s are attached to pods and none of them have a scc with fsgroup,
+  don't use any fsgroup. This will adopt the namespaces default security context.
+
+* If not attached to any pod:
+  * Run without explicit fsgroup, adopting default security context.
+
+Along with the fsgroup, we will also have to work with supplemental group ids,
+which is the same concept as fsgroup but for shared storages like NFS and glusterfs.
+
```
Comment (shawn-hurley) [inline enhancements/direct-migration-filesystem-ownership/README.md:109]: There is more here than @alaypatel07 is working on in regards to the destination pod and how to make sure that it is run as none root and keep the same permissions suck that the application pod when migrated will still be able to read/write.

I think we should add that info here as well.

File: enhancements/direct-migration-filesystem-ownership/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+## Edge cases
```
Comment (shawn-hurley) [inline enhancements/direct-migration-filesystem-ownership/README.md:110]: @alaypatel07 are there potential issues with the destination rysnc server being root with the escape hatch and the application pod having new GIDs/UID

**Review by shawn-hurley — CHANGES_REQUESTED** [review]
We have found more edge cases with the destination that we should continue to add to this doc. All the source cluster bits look right to me, though.

File: enhancements/direct-migration-filesystem-ownership/README.md (hunk 1 above)
Comment (eriknelson) [reply enhancements/direct-migration-filesystem-ownership/README.md:95]: Do we need the selinux context value as well?

Comment (eriknelson) [reply enhancements/direct-migration-filesystem-ownership/README.md:95]: Actually, have we audited the rest of the `PodSecurityContext` to determine all of the fields we actually care about? (https://docs.openshift.com/container-platform/4.6/rest_api/objects/index.html#podsecuritycontext-core-v1). I'd expect that we only care about those that will grant us the correct filesystem privileges.

I'm seeing the `SELinuxOptions` option as a struct itself, I'm assuming we would need to clone this full object?

File: enhancements/direct-migration-filesystem-ownership/README.md (hunk 2 above)
Comment (eriknelson) [reply enhancements/direct-migration-filesystem-ownership/README.md:109]: @alaypatel07 Can you elaborate on the above?

File: enhancements/direct-migration-filesystem-ownership/README.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+I'm going to use `securityContext` in the following as a shorthand for
+`securityContext.{fsGroup,supplementalGroups,seLinuxOptions}`
+
+* If any pvc is attached to one or more pods:
+  * Attached to one pod, get its `securityContext` and use the same `securityContext` for the rsync
+  client pods and explicitly set the node to match the current pod that the
+  volume is attached to.
+  * Attached to more than one pod, raise a warning because this could lead to
+  indeterminate behavior. Pick any of the pods’s fsGroup at random and use it
```
Comment (shawn-hurley) [inline enhancements/direct-migration-filesystem-ownership/README.md:109]: nit: s/fsGroup/securityContext/ I think

**Review by shawn-hurley — APPROVED** [review]

**Pull request merged**